openlineage, common.sql:  provide OL SQL parser as internal OpenLineage provider API

JDarDagran opened the pull request:
This PR adds OpenLineage SQL parser as internal provider API.
It also adds support for `SQLExecuteQueryOperator`. A set of selected SQL providers will be implemented in next PRs.

closes: #29673

JDarDagran committed:
Add SQLParser class serving as the API for openlineage_sql library.
Implement base methods for SQLExecuteQueryOperator & DbApiHook.

Signed-off-by: Jakub Dardzinski <kuba0221@gmail.com>

Rename methods to expose their purpose for OpenLineage.

Signed-off-by: Jakub Dardzinski <kuba0221@gmail.com>

ashb commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -517,3 +518,30 @@ def test_connection(self):
+        """Method used for SQL parsing. Naively tries to use Connection's conn_type"""
+        return connection.conn_type
+
+    def get_authority(self, connection):
```
Is this envisaged/meant to be public or not?

```suggestion
    def _get_authority(self, connection):
```

JDarDagran replied:
Done.

ashb commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -290,6 +290,32 @@ def prepare_template(self) -> None:
+            return None
+
+        sql_parser = SQLParser(
+            dialect=hook.get_database_dialect(connection), default_schema=hook.get_default_schema()
```
Worth putting this in a try/except block so it doesn't fail if OL doesn't know about the particular conn dialect?

JDarDagran replied:
Done.

ashb commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -290,6 +290,32 @@ def prepare_template(self) -> None:
+        )
+
+        operator_lineage: OperatorLineage = sql_parser.generate_openlineage_metadata_from_sql(
+            sql=self.sql, hook=hook, database_info=database_info, database=self.database
```
Ditto here -- not sure if this could throw?

JDarDagran replied:
The logic for catching exceptions in `get_openlineage_facets_*` was introduced by @mobuchowski in different PR. I've applied the same patch of code.

ashb commented on airflow/providers/openlineage/utils/sqlparser.py:
Personal nit/bugbear of mine: I don't like `utils/` folders as they become a dumping ground for all manner of unrelated fields.

Could this perhaps be `airflow/providers/openlineage/sqlparser.py` instead?

JDarDagran replied:
> Could this perhaps be airflow/providers/openlineage/sqlparser.py instead?

👍 changed.

ashb commented on airflow/providers/openlineage/utils/sqlparser.py:
```diff
@@ -0,0 +1,263 @@
+
+from typing import TYPE_CHECKING, Callable, Iterable
+
+from attr import define
```
`attrs` is the recommend name to use now

```suggestion
from attrs import define
```

JDarDagran replied:
Fixed.

ashb commented on airflow/providers/openlineage/utils/sqlparser.py:
```diff
@@ -0,0 +1,263 @@
+    authority: str | None = None
+    database: str | None = None
+    information_schema_columns: list[str] = DEFAULT_INFORMATION_SCHEMA_COLUMNS
+    information_schema_table_name: str = DEFAULT_INFORMATION_SCHEMA_TABLE_NAME
```
Is this generic enough to be a worthwhile default, or would making it a required field make more sense?

JDarDagran replied:
I'd say both are valuable default values that are common for lots of SQL databases.

ashb commented on airflow/providers/openlineage/utils/sqlparser.py:
```diff
@@ -0,0 +1,263 @@
+        """Makes sure to return a semicolon-separated SQL statements"""
+        if isinstance(sql, str):
+            sql = [stmt for stmt in sql.split(";") if stmt != ""]
+        sql = [obj for stmt in sql for obj in stmt.split(";") if obj != ""]
```
This won't work for pg/PSQL's multiline blocks:

```sql
CREATE FUNCTION somefunc() RETURNS integer AS $$
BEGIN
    ...
END;
$$ LANGUAGE plpgsql
```

(Can be fixed in another PR as I'm guess this is porting from OL Airflow plugin?)

JDarDagran replied:
Fixed with an optional import from `DbApiHook.split_sql_string`, otherwise runs same logic.

ashb commented on generated/provider_dependencies.json:
```diff
@@ -231,7 +231,9 @@
     ],
-    "cross-providers-deps": []
+    "cross-providers-deps": [
+      "openlineage"
```
Does this make it a required dep or optional?

JDarDagran replied:
It makes it optional as extra package for `pip install`

eladkal replied:
I think we need to revert this PR.

This change add `openlineage` extra to `common.sql` but  `openlineage` has never been released. We can not add such extra till we release `openlineage`

mobuchowski replied:
@eladkal So, I think we have chicken and egg problem - the OL provider is not released yet, but it's useless to release without any provider support. Can we not have those changes - which do not change what `common.sql` does, just add methods to it - without releasing OL provider first?

eladkal replied:
We often have cases where providers co-depended. In such senario we release the providers in the same wave so users gain the functionality together.

So this PR should have been put on hold till OL provider is ready then merge them in a sequence so both are ready for the same release wave. In providers release cycle we release from main branch we do not cherry pick commits thus anything merged will be released.


Maybe @potiuk have another idea?

ashb reviewed (commented):
(I haven't looked at the tests in details yet)

mobuchowski commented on airflow/providers/openlineage/extractors/base.py:
```diff
@@ -35,6 +37,15 @@ class OperatorLineage:
+        return OperatorLineage(
+            inputs=base.inputs + other.inputs,
+            outputs=base.outputs + other.outputs,
+            run_facets=merge_dicts(base.run_facets, other.run_facets),
```
I would indicate here that in the future we might need a more sophisticated solution

JDarDagran replied:
Added description.

mobuchowski commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -290,6 +290,51 @@ def prepare_template(self) -> None:
+
+        hook: DbApiHook = self.get_db_hook()
+
+        connection = hook.get_connection(getattr(hook, cast(str, hook.conn_name_attr)))
```
Can we "save" some of those objects on `execute` so we can just refer to them as `self.hook` for example?

It might not matter now, but probably we'd want to make sure we're using the same hook instance.

mobuchowski commented on airflow/providers/openlineage/sqlparser.py:
```diff
@@ -0,0 +1,297 @@
+        """
+        split_statement: Callable[[str], list[str]]
+        try:
+            from airflow.providers.common.sql.hooks.sql import DbApiHook
```
I think SQL provider is one we can hard require. Also, will we ever get called here if SQL provider does not exist?

JDarDagran replied:
It's hard to predict where else SQLParser could be used but I can imagine that could happen without use of `common.sql` provider. That's why I've created the logic to reflect the same behaviour even when no `common.sql` provider is available.

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -516,3 +517,40 @@ def test_connection(self):
+        """
+        return self.__schema or "public"
+
+    def get_database_specific_lineage(self, task_instance):
```
Would it make sense to use the following, in this method's definition?
```
raise NotImplementedError()
```

JDarDagran replied:
It would be better to have an implicitly empty method instead of requiring each hook to explicitly declare it as empty. Most of the hooks won't have any db specific lineage.

tatiana commented on airflow/providers/openlineage/utils/sql.py:
```diff
@@ -0,0 +1,194 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Why do we have the `utils` folder and didn't add `sql` to the provider root? If we need a folder, would there be more specific names?

JDarDagran replied:
I'd say there are at least 2 kinds of utility files now, which is enough reason for me to have a separate folder for it. Probably some more would show up in near future, e.g. when developing additional support PythonOperator and TaskFlow API.

tatiana commented on airflow/providers/openlineage/utils/sql.py:
```diff
@@ -0,0 +1,194 @@
+# (ex: int4, timestamp, varchar, etc)
+_UDT_NAME = 4
+# Database is optional as 5th column
+_TABLE_DATABASE = 5
```
It may be worth to group these constants in an `IntEnum`, since they are related:
```
from enum import IntEnum


class Index(IntEnum):
    SCHEMA = 0
    TABLE = 1
    COLUMN_NAME = 2
    ORDINAL_POSITION = 3
    UDT_NAME = 4
    DATABASE = 5
```

JDarDagran replied:
That's a great idea, thanks!

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,40 @@ def test_connection(self):
 
         return status, message
+
+    def get_database_info(self, connection):
```
Is this specific to lineage? If yes, would it make sense to mention this in the method name explicitly?

It is probably worth adding a docstring.

JDarDagran replied:
It is and I'll add more descriptive docstring.
Also I noticed it should not pass any database info by default implicitly rather than have each provider hook implement it. Will do the changes.

tatiana replied:
To be renamed get_database_info -> get_openlineage_database_info

This method should be ` :meta private:`
https://github.com/apache/airflow/blob/b7796895cb41d8e5e79e6d8eee150b11d8c302a7/airflow/utils/mixins.py#L57-L63

JDarDagran replied:
Does `:meta private:` make method to be skipped in docs generation?
If so we probably should follow this rule for all OpenLineage related methods or none of them.

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,40 @@ def test_connection(self):
+            scheme=self.get_database_dialect(connection), authority=self._get_authority(connection)
+        )
+
+    def get_database_dialect(self, connection):
```
In which scenarios the database dialect isn't the same as `connection.conn_type`?

JDarDagran replied:
I'd assume it might be different in any provider for some reason. Dialect is specific to OpenLineage SQLParser and accepts only a limited number of dialects. A vast of `conn_type`s should then return `generic` or `postgres` - e.g. `redshift` one.

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,40 @@ def test_connection(self):
+        """Method used for SQL parsing. Naively tries to use Connection's conn_type"""
+        return connection.conn_type
+
+    def _get_authority(self, connection):
```
Since this is a method specific to lineage / openlineage, it may be worth to add this to the method name.

JDarDagran replied:
:+1:

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,40 @@ def test_connection(self):
+        return connection.conn_type
+
+    def _get_authority(self, connection):
+        """Returns authority part (without user info) of OpenLineage namespace URI."""
```
Perhaps the docstring could contain an example and focus on what it returns (host and port), as opposed to what it doesn't (path)

JDarDagran replied:
:+1:

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,40 @@ def test_connection(self):
+
+    def get_database_dialect(self, connection):
+        """Method used for SQL parsing. Naively tries to use Connection's conn_type"""
+        return connection.conn_type
```
Generic should be the default and we should define it at the top of the module

JDarDagran replied:
It is set "generic" but for now I don't see a particular reason to define this as a variable at the top of the module.

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,40 @@ def test_connection(self):
+        Returns default schema specific to database.
+        See: :class:`~providers.openlineage.utils.sqlparser.SQLParser`
+        """
+        return self.__schema or "public"
```
In the case of Postgres, `self.__schema` will return the `database` and not the schema.
We need to find a way of allowing users to override this.

JDarDagran replied:
Postgres in an exception and this would be changed in PostgresHook separately.

sunank200 commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -290,6 +291,64 @@ def prepare_template(self) -> None:
+        except ImportError:
+            return None
+
+        hook: DbApiHook = self.get_db_hook()
```
This is the different hook instance. Why not use the same instance of hook which execute method uses?

JDarDagran replied:
It returns `self._hook` which is decorated as `cached_property`. What would be the other reasonable way to use the same instance?

uranusjr replied:
Just curious, why is the `DbApiHook` annotation needed? (Same question applies to a lot of code in this PR really)

JDarDagran replied:
Removed this annotation as well as some others unnecessary in this PR.

sunank200 commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -290,6 +291,64 @@ def prepare_template(self) -> None:
+        except ImportError:
+            return operator_lineage
+
+        hook: DbApiHook = self.get_db_hook()
```
This is the different hook instance. Why not use the same instance of hook which execute method uses?

JDarDagran replied:
Same as in another comment:

> It returns self._hook which is decorated as cached_property. What would be the other reasonable way to use the same instance?

sunank200 reviewed (commented):
Added few comments

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,55 @@ def test_connection(self):
+        :return: A :class:`airflow.providers.openlineage.sqlparser.DatabaseInfo` instance.
+        """
+
+    def get_database_dialect(self, connection) -> str:
```
What do you think if suffixing all open lineage-specific methods with `openlineage`?

JDarDagran replied:
Maybe not suffix but like `get_openlineage_whatever_goes_next`?

uranusjr replied:
I like putting `openlineage` after `get_`; reads better than a suffix.

JDarDagran replied:
Done.

tatiana commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,55 @@ def test_connection(self):
+        """
+        return "generic"
+
+    def get_default_schema(self):
```
What do you think if suffixing all open lineage-specific methods with `openlineage`?

JDarDagran replied:
Done.

uranusjr commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,55 @@ def test_connection(self):
+        and conforms OpenLineage naming convention for a number of databases (e.g. MySQL, Postgres, Trino).
+        """
+        if connection.port and connection.host:
+            authority = f"{connection.host}:{connection.port}"
```
Is this block necessary? I thought the `get_uri` approach would produce the same result for this case.

JDarDagran replied:
`get_uri` returns a full URI so that e.g. `sqlalchemy` can connect with it. Most importantly it contains login and password which should not definitely be passed to OpenLineage event. Generally speaking for those mentioned databases in docstring we should just capture host and port.

uranusjr replied:
Yes but in the else block below you can parse out the host and port from the URI. Can this not be done for all cases? Why is the `connection.port and connection.host` case special?

JDarDagran replied:
I think you're right. We can stick to what's in `else` block.
Changed this.

pankajastro reviewed (approved):
LGTM

phanikumv commented on airflow/providers/common/sql/hooks/sql.py:
```diff
@@ -515,3 +516,52 @@ def test_connection(self):
+        and creating correct namespace.
+
+        :param connection: Airflow connection to reduce calls of `get_connection` method
+        :return: A :class:`airflow.providers.openlineage.sqlparser.DatabaseInfo` instance.
```
```suggestion
```

JDarDagran replied:
Why so?

uranusjr replied:
This can be expressed in the type annotation instead (so you need to add a `-> DatabaseInfo` if this is deleted)

JDarDagran replied:
As it's empty method that may be optionally implemented, is it enough to annotate it with `-> DatabaseInfo | None`? Does this say enough? I'm no expert in static typing.

JDarDagran committed:
Rewrite information schema query construction to SQLALchemy ORM.

Signed-off-by: Jakub Dardzinski <kuba0221@gmail.com>

uranusjr commented on airflow/providers/openlineage/utils/sql.py:
```diff
@@ -0,0 +1,199 @@
+    filter_clauses = []
+    for schema, tables in schema_mapping.items():
+        filter_clause = information_schema_table.c.table_name.in_(
+            map(lambda name: name.upper() if uppercase_names else name, tables)
```
```suggestion
            name.upper() if uppercase_names else name for name in tables
```

Seems shorter and easier to understand to me.

JDarDagran replied:
Done.

uranusjr commented on airflow/providers/openlineage/extractors/base.py:
```diff
@@ -88,16 +89,29 @@ def extract(self) -> OperatorLineage | None:
+                outputs=facets.outputs,
+                run_facets=facets.run_facets,
+                job_facets=facets.job_facets,
+            )
```
Does this line generate errors caught below? If not, it would be better to put this in an `else` block.

JDarDagran replied:
Put into `else` block.

uranusjr committed:
Clean up in-class reference

Instead of referencing the SQLParser directly, modify various static
methods to class methods instead, so they can use the cls argument
to avoid spelling out the class name repeatedly.

Also added a few changes to better ultilize type reference and eliminate
some verbose type annotations.

uranusjr commented on airflow/providers/openlineage/utils/sql.py:
```diff
@@ -0,0 +1,195 @@
+        return Dataset(
+            namespace=namespace,
+            name=name,
+            facets={"schema": SchemaDatasetFacet(fields=self.fields)} if len(self.fields) is not None else {},
```
The `len()` part can never be None and thus the `else` part can never execute. Is there a typo here?

JDarDagran replied:
It is, fixed.

uranusjr committed:
Clean up typing and iterator usage

JDarDagran committed:
Add static typing to hint returned type.

Signed-off-by: Jakub Dardzinski <kuba0221@gmail.com>

JDarDagran committed:
Fix mypy issues.

Signed-off-by: Jakub Dardzinski <kuba0221@gmail.com>
